### PR TITLE
Fix capture promotion and closure scope analysis' handling of dynamically_replaceable functions 

### PR DESCRIPTION
**Pull request by aschwaighofer**
rdar://50949761

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: lib/SILOptimizer/Analysis/ClosureScope.cpp [hunk 1]
```diff
@@ -82,7 +82,8 @@ class ClosureScopeData {
   void recordScope(PartialApplyInst *PAI) {
     // Only track scopes of non-escaping closures.
     auto closureTy = PAI->getCallee()->getType().castTo<SILFunctionType>();
-    if (!isNonEscapingClosure(closureTy))
+    if (!isNonEscapingClosure(closureTy) ||
```
Comment (atrick) [inline lib/SILOptimizer/Analysis/ClosureScope.cpp]: It doesn't really make sense to me that both isNonEscaping and isCalleeDynamicallyReplaceable could be true.
For future reference, we should have a comment explaining how this came to pass so we can fix it with a proper assert later.

File: lib/SILOptimizer/IPO/CapturePromotion.cpp [hunk 2]
```diff
@@ -977,7 +977,7 @@ bool isPartialApplyNonEscapingUser(Operand *CurrentOp, PartialApplyInst *PAI,
   unsigned Index = OpNo - 1 + closureConv.getNumSILArguments();
 
   auto *Fn = PAI->getReferencedFunction();
-  if (!Fn || !Fn->isDefinition()) {
+  if (!Fn || !Fn->isDefinition() || Fn->isDynamicallyReplaceable()) {
```
Comment (atrick) [inline lib/SILOptimizer/IPO/CapturePromotion.cpp]: Please comment explaining why it this pass in particular cannot safe look at the dynamically replaceable body.

**Review by atrick — APPROVED** [review]
I don't like the fact that a SIL pass can ask for an Apply/PartialApply's referenced SILFunction and get back a dynamically referenced function. I can't think of a more dangerous and misleading SIL API. Can you create the necessary FIXMEs and bugs to make sure we don't leave it in this state?

**Comment by aschwaighofer** [comment]
https://bugs.swift.org/browse/SR-10720 Fix getReferencedFunction API for dynamically_replaceable SIL functions

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please Test linux

**Pull request merged**